File: Editorial/0114_Flatten_Binary_Tree_to_Linked_List/CSharp/2_Iterative_Solution_using_Stack.cs
public class Solution {
    private (TreeNode, int) GetPair(TreeNode n, int v) {
        return (n, v);
    }

    public void Flatten(TreeNode root) {
        if (root == null) {
            return;
        }

        int START = 1, END = 2;
        TreeNode tail = null;
        Stack<(TreeNode, int)> stack = new Stack<(TreeNode, int)>();
        stack.Push(GetPair(root, START));
        while (stack.Count > 0) {
            var nodeData = stack.Pop();
            TreeNode node = nodeData.Item1;
            int state = nodeData.Item2;
            if (node.left == null && node.right == null) {
                tail = node;
                continue;
            }

            if (state == START) {
                if (node.left != null) {
                    stack.Push(GetPair(node, END));
                    stack.Push(GetPair(node.left, START));
                } else if (node.right != null) {
                    stack.Push(GetPair(node.right, START));
                }
            } else {
                TreeNode rightNode = node.right;
                if (tail != null) {
                    tail.right = node.right;
                    node.right = node.left;
                    node.left = null;
                    rightNode = tail.right;
                }

                if (rightNode != null) {
                    stack.Push(GetPair(rightNode, START));
                }
            }
        }
    }
}

File: Editorial/0114_Flatten_Binary_Tree_to_Linked_List/TypeScript/2_Iterative_Solution_using_Stack.ts
function flatten(root: TreeNode | null): void {
    if (root === null) {
        return;
    }
    let tailNode: TreeNode | null = null;
    const START = 1,
        END = 2;
    let stack: Array<[TreeNode, number]> = [[root, START]];
    while (stack.length > 0) {
        let nodeData = stack.pop() as [TreeNode, number];
        let currentNode = nodeData[0];
        let recursionState = nodeData[1];
        if (currentNode.left === null && currentNode.right === null) {
            tailNode = currentNode;
            continue;
        }
        if (recursionState === START) {
            if (currentNode.left !== null) {
                stack.push([currentNode, END]);
                stack.push([currentNode.left, START]);
            } else if (currentNode.right !== null) {
                stack.push([currentNode.right, START]);
            }
        } else {
            let rightNode = currentNode.right;
            if (tailNode !== null) {
                tailNode.right = currentNode.right;
                currentNode.right = currentNode.left;
                currentNode.left = null;
                rightNode = tailNode.right;
            }
            if (rightNode !== null) {
                stack.push([rightNode, START]);
            }
        }
    }
}
